File: 448.java
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        // time: O(n), space: O(1)
        List<Integer> res = new ArrayList<>();
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            int idx = Math.abs(nums[i]) - 1;
            nums[idx] = -Math.abs(nums[idx]);
        }
        for (int i = 0; i < n; i++) {
            if (nums[i] > 0) {
                res.add(i + 1);
            }
        }
        return res;
    }
}

File: 448.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        # time: O(n), space: O(1)
        for n in nums:
            while nums[n - 1] != n:
                nums[n - 1], n = n, nums[n - 1]
        return [i + 1 for i in range(len(nums)) if nums[i] != i + 1]

        # time: O(n), space: O(n)
        # return list(set(x for x in range(1, len(nums) + 1)) - set(nums))